### PR TITLE
SONARCS-602 Support the SARIF format used by VS2015 Update 3

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+      switch (version) {
+        case "0.4":
+        case "0.1":
+          return new SarifParser01And04(contents);
+        case "1.0.0":
+        case "1.0":
+          return new SarifParser10(contents);
+        default:
+          throw new IllegalStateException(String.format("Unable to parse the Roslyn SARIF report file: %s. Unrecognized version: %s", file.getAbsolutePath(), version));
```
Comment (henryju) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java]: This is an open question here: should we fail or rely on the fact format will be backward compatible from now on and use Parser10?

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParser01And04.java [hunk 2]
```diff
@@ -17,59 +17,48 @@
...
 
-  public void parse(File file) {
-    String contents;
-    try {
-      contents = Files.toString(file, Charsets.UTF_8);
-    } catch (IOException e) {
-      throw new IllegalStateException("Unable to read the Roslyn SARIF report file: " + file.getAbsolutePath(), e);
-    }
-
+  public void parse(SarifParserCallback callback) {
```
Comment (SonarTech) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParser01And04.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1161)

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParser01And04.java [hunk 3]
```diff
@@ -17,59 +17,48 @@
...
+  private void handleIssues(JsonArray issues, boolean linesOffByOne, SarifParserCallback callback) {
     for (JsonElement issueElement : issues) {
       JsonObject issue = issueElement.getAsJsonObject();
-      handleIssue(issue, linesOffByOne);
+      handleIssue(issue, linesOffByOne, callback);
     }
   }
 
-  private void handleIssue(JsonObject issue, boolean linesOffByOne) {
+  private void handleIssue(JsonObject issue, boolean linesOffByOne, SarifParserCallback callback) {
```
Comment (SonarTech) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParser01And04.java]: MAJOR Make "handleIssue" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParser.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.csharp.sarif;
+
+public interface SarifParser {
+  public void parse(SarifParserCallback callback);
```
Comment (SonarTech) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParser.java]: MINOR "public" is redundant in this context. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2333)

File: src/main/java/org/sonar/plugins/csharp/CSharpSensor.java [hunk 5]
```diff
@@ -67,6 +67,10 @@
 import org.sonar.api.utils.command.Command;
 import org.sonar.api.utils.command.CommandExecutor;
 import org.sonar.api.utils.command.StreamConsumer;
+import org.sonar.plugins.csharp.sarif.SarifParser;
+import org.sonar.plugins.csharp.sarif.SarifParser01And04;
```
Comment (SonarTech) [inline src/main/java/org/sonar/plugins/csharp/CSharpSensor.java]: MINOR Remove this unused import 'org.sonar.plugins.csharp.sarif.SarifParser01And04'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParser10.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+public class SarifParser10 implements SarifParser {
+  private static final String FILE_PROTOCOL = "file:///";
+  private String contents;
+
+  SarifParser10(String contents) {
+    this.contents = contents;
+  }
+
+  public void parse(SarifParserCallback callback) {
```
Comment (SonarTech) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParser10.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1161)

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParser10.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  private void handleIssues(JsonArray results, SarifParserCallback callback) {
+    for (JsonElement resultEl : results) {
+      JsonObject resultObj = resultEl.getAsJsonObject();
+      handleIssue(resultObj, callback);
+    }
+  }
+
+  private void handleIssue(JsonObject resultObj, SarifParserCallback callback) {
```
Comment (SonarTech) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParser10.java]: MAJOR Make "handleIssue" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.io.File;
+import java.io.IOException;
+
+import com.google.common.base.Charsets;
+import com.google.common.io.Files;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+public class SarifParserFactory {
```
Comment (SonarTech) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1118)

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  public static SarifParser create(File file) {
```
Comment (SonarTech) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java]: MAJOR The Cyclomatic Complexity of this method "create" is 12 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues
* MAJOR 8 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [sarif](https://github.com/SonarSource/sonar-csharp/blob/afdf2400b109d94103aed9e816e777266f8ed73c/src/main/java/org/sonar/plugins/csharp/sarif): Add a 'package-info.java' file to document the 'sarif' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)
1. MAJOR [SarifParser01And04.java#L82](https://github.com/SonarSource/sonar-csharp/blob/afdf2400b109d94103aed9e816e777266f8ed73c/src/main/java/org/sonar/plugins/csharp/sarif/SarifParser01And04.java#L82): Add parentheses to make the operator precedence explicit. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS864)

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java (hunk 1 above)
Comment (tvsonar) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java]: @henryju That's true, but it shouldn't block us for the time being. We can make this decision later, after the bugfix release.

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+import com.google.common.io.Files;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+public class SarifParserFactory {
+  private SarifParserFactory() {
+    // private
+  }
+
+  public static SarifParser create(File file) {
```
Comment (SonarTech) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java]: MAJOR The Cyclomatic Complexity of this method "create" is 12 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java (hunk 1 above)
Comment (dbmeneses) [inline src/main/java/org/sonar/plugins/csharp/sarif/SarifParserFactory.java]: I will change it to always use Parser 1.0 when version is not known, and we'll think again what's the best way to do it once we have a answer from Jean Marc and implement SONARCS-603

**Comment by henryju** [comment]
LGTM

**Pull request merged**